### PR TITLE
Offers: Part 1, Spec import and basic infrastructure

### PR DESCRIPTION
**Pull request by rustyrussell**
BOLT 12 import, and various helpers to support that (offers, invoice_request, invoice) including signing infrastructure.

Almost all is EXPERIMENTAL_FEATURES, though we'll probably move to a runtime option before release.

Changelog-None

### CODE REVIEW

**Comment by PapauloGamerOfc** [comment]
The Travis CI bot is failing on every PR!

File: common/bolt12.c [hunk 1]
```diff
@@ -0,0 +1,301 @@
...
+	 * - if the chain for the invoice is not solely bitcoin:
+	 *   - MUST specify `chains` the invoice is valid for.
+	 * - otherwise:
+	 *   - the bitcoin chain is implied as the first and only entry.
+	 */
+	num_chains = tal_count(chains);
+	if (num_chains == 0) {
+		num_chains = 1;
+		chains = &chainparams_for_network("bitcoin")->genesis_blockhash;
+	}
```
Comment (cdecker) [inline common/bolt12.c:41]: Could this lead to a leak?

File: common/bolt12.c [hunk 2]
```diff
@@ -0,0 +1,301 @@
...
+{
+	if (must_be_chain) {
+		if (!bolt12_chains_match(chains, must_be_chain))
+			return tal_fmt(ctx, "wrong chain");
+	}
+
+	if (our_features) {
+		int badf;
+
+		badf = features_unsupported(our_features, features,
```
Comment (cdecker) [inline common/bolt12.c:66]: Nit: merge these lines or move `int badf` to top.

**Review by cdecker — APPROVED** [review]
Very nice PR, just some minor clarifications :-)

ACK https://github.com/ElementsProject/lightning/pull/4255/commits/40217011b980f281039d9fb33f5573fb8fcb75b2

File: bitcoin/signature.c [hunk 3]
```diff
@@ -348,3 +348,15 @@ void towire_bitcoin_signature(u8 **pptr, const struct bitcoin_signature *sig)
 	towire_secp256k1_ecdsa_signature(pptr, &sig->s);
 	towire_u8(pptr, sig->sighash_type);
 }
+
+void towire_bip340sig(u8 **pptr, const struct bip340sig *bip340sig)
```
[the remark this answers is not in the document]
Comment (rustyrussell) [reply bitcoin/signature.c:352]: It's the name I used in the spec.  I can change both, but we called the previous type `signature` so I need something distinct from that.

File: common/bolt12.c (hunk 1 above)
Comment (rustyrussell) [reply common/bolt12.c:41]: No, it points into a static array.

**Comment by rustyrussell** [comment]
Trivial change to make `int badf; badf = ...` into `int badf = ...`.

**Pull request merged**